src/libos/crates/sgx-untrusted-alloc/src/untrusted_allocator/mod.rs

```diff
@@ -0,0 +1,101 @@
+use super::*;
+
+use crate::prelude::*;
+
+mod free_space_manager;
+mod vm_area;
+mod vm_chunk_manager;
+mod vm_range;
+mod vm_util;
+
+const PAGE_SIZE: usize = 4096;
+
+// The address of a block returned by malloc or realloc in GNU systems is always a multiple of eight (or sixteen on 64-bit systems).
+const DEFAULT_ALIGNMENT: usize = 16;
+
+use vm_range::*;
+use vm_util::*;
+
+use libc::c_void;
+use libc::{MAP_ANONYMOUS, MAP_PRIVATE, PROT_READ, PROT_WRITE};
+use spin::Mutex;
+use vm_chunk_manager::ChunkManager as VMManager;
+cfg_if::cfg_if! {
+    if #[cfg(feature = "sgx")] {
+        use libc::ocall::{mmap, munmap};
+    } else {
+        use libc::{mmap, munmap};
+    }
+}
+
+pub struct Allocator {
+    range: VMRange,
+    inner: Mutex<VMManager>,
+}
+
+impl Allocator {
+    // Initiate a untrusted memory allocator with size specified by user.
+    // This function is marked "unsafe" because this allocator will allocate
+    // untrusted memory which could lead to much unsafe behavior.
+    // The user must use untrusted memory with more caution.
+    pub unsafe fn new(size: usize) -> Self {
```

Comment:
I got you intention. You meant well. But still, this method should not be marked as unsafe.

Allow me to borrow an explanation from StackOverflow.

> Mark a function as `unsafe` iff the function's safety depends on its parameters or on global state. If the function is safe regardless of arguments and global state, don't mark it as `unsafe`. Whether you consider a function that uses `unsafe` internally safe is the same as whether you consider a C program safe.

You see. What `unsafe` says is: calling this function is only safe on some conditions.

In our situation, calling the `new` method is safe and secure _unconditionally_, which just creates the allocator itself. You can create an infinite number of allocators without causing any safety or security problem. So the `new` method is ok. It is safe.

What may be unsafe or insecure is using the `alloc` method, only through which the user can use untrusted memory. That is the direct cause of any potential safety or security problem. So that is the method that should  be marked unsafe.

Reply:
Got it. Thanks for your review. Done.